### PR TITLE
Small cleanups around the ssr implementation of applyn

### PR DESCRIPTION
**Pull request by ppedrot**
From the series: cleanups from my attic.

We eliminate dead code and uniformize evar generation w.r.t. subtle and most likely irrelevant details inherited from historical reasons.

### CODE REVIEW

**Comment by gares** [comment]
I'm a bit bluffed by the invariant that applyn is only called with terms having explicit lambdas.
Since it seems true, then its name/comment should say it.
I think the old code was not assuming that.

**Comment by ppedrot** [comment]
The unicoq failure is real, it's due to the commit introducing the unconditional β reduction. I'm not sure what is going on but we're feeding an ill-typed term to the kernel after this patch.

**Comment by ppedrot** [comment]
For backwards compatibility I have removed the patch that performed an unconditional β reduction.

**Comment by gares** [comment]
> I'm a bit bluffed by the invariant that applyn is only called with terms having explicit lambdas.

Can you explain this?

**Comment by ppedrot** [comment]
> Then it should be written somewhere!!

I've added a comment already, but rather than writing it it should be enforced by the type system, returning a term in a context rather than just the expanded body. I have a few more cleanups around this piece of code but didn't consider submitting them as they're not ready. I thought it was better to proceed cautiously (as the CI actually caught a mysterious regression with beta-reduced terms).

File: plugins/ssr/ssrcommon.ml [hunk 1]
```diff
@@ -811,39 +803,54 @@ let dependent_apply_error =
  * Refine.refine function that handles type classes and evars but fails to
  * handle "dependently typed higher order evars".
  *
+ * Assumes that the type of [t] is a telescope with [n] leading product types.
+ * This is always the case as the only caller of applyn uses it with the output
+ * of abs_evars_pirrel which starts with n lambdas generated by abstraction.
+ *
```
Comment (gares) [inline plugins/ssr/ssrcommon.ml:809]:
```suggestion
 * Assumes that the type of [t] is a telescope with [n] leading product types.
 * This is always the case as the only caller of applyn uses it with the output
 * of abs_evars_pirrel which starts with n lambdas generated by abstraction.
 * It also assumes [t] starts with [n] Lambda nodes.
 *
```

Comment (gares) [reply plugins/ssr/ssrcommon.ml:809]: This is what I meant.

Comment (ppedrot) [reply plugins/ssr/ssrcommon.ml]: Done, in any case.

**Comment by gares** [comment]
Tell me when it is ready, there is still a failure

**Comment by ppedrot** [comment]
I think that the failure is from the previous version with the unconditional beta reduction, now the CI seems fine (fiat-crypto failure is an unrelated timeout, as usual).

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
@coqbot merge now

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**